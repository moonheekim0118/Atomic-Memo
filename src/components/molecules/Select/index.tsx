import React, { useEffect, useState, useCallback } from 'react';
import useToggle from '../../../hooks/useToggle';
import Option from '../../atoms/Option';
import { memoList } from '../../../util/memoKind';
import styled from 'styled-components';

const DEFAULT_TEXT = 'Select Kind of Memo!';

const Select = () => {
  const [defaultText, setDefaultText] = useState<string>(DEFAULT_TEXT);
  const [showList, toggleList, setShowList] = useToggle(false);

  useEffect(() => {
    // 외부 클릭시 List 안보이도록
    function onClickOutSide(this: Document, ev: MouseEvent) {
      const target = ev.target as HTMLTextAreaElement;
      const classList = target.classList;
      if (
        !classList.contains('select-option') &&
        !classList.contains('selected-text')
      ) {
        setShowList(false);
      }
    }

    document.addEventListener('mousedown', onClickOutSide);

    return () => {
      document.removeEventListener('mousedown', onClickOutSide);
    };
  }, []);

  // 옵션 클릭시
  const onClickOption = useCallback((e: React.MouseEvent) => {
    e.stopPropagation();
    const target = e.target as HTMLTextAreaElement;
    setDefaultText(target.getAttribute('data-name'));
    setShowList(false);
  }, []);

  return (
    <Container>
      <DefaultText
        className={showList ? 'selected-text active' : 'selected-text'}
        onClick={toggleList}>
        {defaultText}
      </DefaultText>
      {showList && (
        <SelectOptions className="select-options" onClick={onClickOption}>
          {memoList.map((v, i) => (
            <Option key={i} dataName={v}>
              {v}
            </Option>
          ))}
        </SelectOptions>
      )}
    </Container>
  );
};

const Container = styled.article`
  cursor: pointer;
  width: 200px;
  position: relative;
  background-color: #3399ff;
  color: #fff;

  padding: 0.7em;
`;

const DefaultText = styled.div`
  ::after {
    content: '';
    position: absolute;
    right: 5px;
    top: 16px;
    border: 7px solid transparent;
    border-color: #fff transparent transparent transparent;
  }

  &.active::after {
    top: 8px;
    border-color: transparent transparent #fff transparent;
  }
`;

const SelectOptions = styled.ul`
  list-style-type: none;
  position: absolute;
  width: 100%;
  left: 0;
  padding-top: 0.5em;
`;

export default Select;
