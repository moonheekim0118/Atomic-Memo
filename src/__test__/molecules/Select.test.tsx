import 'jsdom-global/register';
import React from 'react';
import Select from '../../components/molecules/Select';
import { mount } from 'enzyme';

describe('<Select/>', () => {
  let container = null;
  const mockFn = jest.fn();

  it('renders correctly', () => {
    container = mount(<Select defaultText="Work" setDefaultText={mockFn} />);
  });

  it('matches snpashot', () => {
    expect(container.html()).toMatchSnapshot();
  });
});
