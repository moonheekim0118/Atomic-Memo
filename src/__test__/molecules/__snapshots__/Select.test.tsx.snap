// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Select/> matches snpashot 1`] = `"<article class=\\"Select__Container-sc-1dr9cr-0 kPhMeR\\"><div class=\\"Select__DefaultText-sc-1dr9cr-1 jGhHrZ selected-text\\">Work</div></article>"`;
